Added python notebook for california housing price prediction

swaingotnochill opened the pull request:
It is still in initial stage as the main algorithm implementation has error.

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,726 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "--2021-06-16 00:12:12--  https://matrix.org/_matrix/media/r0/download/matrix.org/WvrgbgzkyIMbvkxLkKKNyMrO/housing.csv\n",
```
@rcurtin can you upload https://matrix.org/_matrix/media/r0/download/matrix.org/WvrgbgzkyIMbvkxLkKKNyMrO/housing.csv to mlpack.org as well.

zoq replied:
Let's add `-q` here to reduce the output.

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,726 @@
+    }
+   ],
+   "source": [
+    "%matplotlib inline\n",
```
Do you mind to put that right below where you import `import matplotlib.pyplot as plt` I think it makes more sense if both are at the same place.

zoq replied:
Looks like this is still something we should do.

swaingotnochill replied:
I am still in fixing a lot of things in terms of style and code. Sure will do.

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,726 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlkAAAGtCAYAAAAlE2HVAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nOy9d3hc1bX3/9lnqkajOurFklzkinvF4E43EEzvBEIJkJB2k3BvnjfJTXlzn3t/Se59k5AQCJjmACFgwgWMKcbYGBsb927LsiyrS1YZSVPP/v2xx7Zsq4zkkWyi/XmeeWZmn7332WdGmvnOWmuvJaSUaDQajUaj0Whii3GuF6DRaDQajUbzz4g ...
```
As @rcurtin mentioned in the chat, I guess it would make sense to put the image into the examples repository, since it's very specific to the notebook and I don't think it's used for anything else.

zoq replied:
I would put this to the top, just to have all the python packages we need for the notebook in one place.

swaingotnochill commented:
@zoq remember in last metting I said something about testLabels changing..Yaah, so the thing is if you see the python notebook dataset.info, you can find that the minimum value for our labels is 14999 while in the testLabels the values are in thousands...I don't know what's causing this, or am i wrong somewhere? please take a look.

zoq commented:
Hm, not sure I can see what you see, I loaded the notebook, and added `arma::min(testLabels)` and `arma::min(testLabels)` right after the split `data::Split(dataset, labels, trainSet, testSet, trainLabels, testLabels, 0.2)` and get `min = 1.0000000` which is `1.0` and `12912.000` which I think is correct.

About the changing values, keep in mind that the split function per default does a random split, so you could end up with different values, you can disable the random split:

https://github.com/mlpack/mlpack/blob/64b9d63f02156e75b79723d485402ba721edea83/src/mlpack/core/data/split_data.hpp#L298-L299

But I'm not sure I missed something or we see the same thing.

swaingotnochill commented:
> Hm, not sure I can see what you see, I loaded the notebook, and added `arma::min(testLabels)` and `arma::min(testLabels)` right after the split `data::Split(dataset, labels, trainSet, testSet, trainLabels, testLabels, 0.2)` and get `min = 1.0000000` which is `1.0` and `12912.000` which I think is correct.
> 
> About the changing values, keep in mind that the split function per default does a random split, so you could end up with different values, you can disable the random split:
> 
> https://github.com/mlpack/mlpack/blob/64b9d63f02156e75b79723d485402ba721edea83/src/mlpack/core/data/split_data.hpp#L298-L299
> 
> But I'm not sure I missed something or we see the same thing.

@zoq  What I was talking is, testLabels is from my median house price of original dataset. In my original dataset, the minimum value is from 14999. So, how come it is different in cpp file. Pardon, I think its not due to split which is assumed, but I still can't get it around when I print the testLabels, the values are in range of thousand when clearly all values should be above 14999.

swaingotnochill commented:
> > Hm, not sure I can see what you see, I loaded the notebook, and added `arma::min(testLabels)` and `arma::min(testLabels)` right after the split `data::Split(dataset, labels, trainSet, testSet, trainLabels, testLabels, 0.2)` and get `min = 1.0000000` which is `1.0` and `12912.000` which I think is correct.
> > About the changing values, keep in mind that the split function per default does a random split, so you could end up with different values, you can disable the random split:
> > https://github.com/mlpack/mlpack/blob/64b9d63f02156e75b79723d485402ba721edea83/src/mlpack/core/data/split_data.hpp#L298-L299
> > But I'm not sure I missed something or we see the same thing.
> 
> @zoq What I was talking is, testLabels is from my median house price of original dataset. In my original dataset, the minimum value is from 14999. So, how come it is different in cpp file. Pardon, I think its not due to split which is assumed, but I still can't get it around when I print the testLabels, the values are in range of thousand when clearly all values should be above 14999.
@zoq issue is sorted...

zoq commented:
What version of mlpack are you testing against? Also what happens if you remove the header and convert `NEAR BAY` to `NEAR-BAY`, does that solve the issue on your side?

swaingotnochill commented:
> What version of mlpack are you testing against? Also what happens if you remove the header and convert `NEAR BAY` to `NEAR-BAY`, does that solve the issue on your side?
Its 3.4.2. The issue was that the csv file was with headers of columns. So, when you load the data along with DatasetInfo, it considered the entire columns of numerical data as categorical. So, If I have to work further, I have to remove the headers first.

zoq commented:
> > What version of mlpack are you testing against? Also what happens if you remove the header and convert `NEAR BAY` to `NEAR-BAY`, does that solve the issue on your side?
> > Its 3.4.2. The issue was that the csv file was with headers of columns. So, when you load the data along with DatasetInfo, it considered the entire columns of numerical data as categorical. So, If I have to work further, I have to remove the headers first.

I see, yes, that issue was fixed in the master branch. You can manually remove the header, no problem from my side.

zoq commented:
@swaingotnochill Do you need any help with the plot conversion from the python notebook?

swaingotnochill commented:
Thanks...I am pretty much done with the visualisation...once I complete the notebook, I will push it.

zoq commented on utils/pandasscatter.py:
```diff
@@ -0,0 +1,28 @@
+    dataset = pd.read_csv(inFile)
+    implot = plt.imshow(im, extent=[-124.55, -113.80, 32.45, 42.05], alpha=0.5,
+           cmap=plt.get_cmap("jet"))
+    plt.scatter(x=dataset[x], y=dataset[y], s=dataset["population"]/100, label=label, c=dataset[c], cmap=plt.get_cmap("jet"), alpha= 0.5)
```
Should we make the dataset column name configurable (`population`), that way users can reuse the function for other datasets as well.

swaingotnochill replied:
I thought about it too...but at the end I have configured 's' according to my needs( its divided by 100)..So, I thought it won't have any effect...Other than that, I don't know how to pass that custom dividend number ( here it is 100). Zoq if you have any ideas regarding it do, tell.

zoq replied:
You could add the divider to the parameters as well:

so instead of using:
`cpandasscattermap(inFile, imgFile, x, y, label, c, outFile="output2.png", height=10, width = 7):`

you can write:

`cpandasscattermap(inFile, imgFile, x, y, label, c, outFile="output2.png", height=10, width=7, divider=100):`

and use that in the code.

swaingotnochill replied:
I tried it, but it's messing up with the plot....The plot looks horrible :(

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,1223 @@
+    "\n",
+    "### Objective\n",
+    "* To predict California Housing Prices using the most simple Linear Regression Model and see how it performs.\n",
+    "* To understand the modeling workflow using MLPACK.\n",
```
```suggestion
    "* To understand the modeling workflow using mlpack.\n",
```

Just to be consistent.

swaingotnochill replied:
My bad :+1:

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,1223 @@
+    "### About the Data\n",
+    " This dataset is a modified version of the California Housing dataset available from Luís Torgo's page (University of Porto). Luís Torgo obtained it from the StatLib repository (which is closed now). The dataset may also be downloaded from StatLib mirrors.\n",
+    " \n",
+    " This dataset is also used in a book HandsOn-ML ( a very good book and highly recommended).\n",
```
```suggestion
    " This dataset is also used in a book HandsOn-ML (a very good book and highly recommended).\n",
```

Remove the extra whitespace, also maybe we should add a link to the book?

swaingotnochill replied:
May be we can add. Will a amazon link will do?

zoq replied:
Let's use https://www.oreilly.com/library/view/hands-on-machine-learning/9781491962282/

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,1223 @@
+    }
+   ],
+   "source": [
+    "# We are using matplotlib for visualization\n",
```
```suggestion
    "# We are using matplotlib for visualization.\n",
```

Use the style defined in https://github.com/mlpack/mlpack/wiki/DesignGuidelines#comments.

swaingotnochill replied:
:+1: Thanks

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,1223 @@
+    }
+   ],
+   "source": [
+    "# let's plot  a scatter plot according to co-ordinates\n",
```
```suggestion
    "# Let's plot  a scatter plot according to coordinates.\n",
```

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,1223 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlkAAAGtCAYAAAAlE2HVAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nOy9d3hc1bX3/9lnqkajOurFklzkinvF4E43EEzvBEIJkJB2k3BvnjfJTXlzn3t/Se59k5AQCJjmACFgwgWMKcbYGBsb927LsiyrS1YZSVPP/v2xx7Zsq4zkkWyi/XmeeWZmn7332WdGmvnOWmuvJaSUaDQajUaj0Whii3GuF6DRaDQajUbzz4g ...
```
```suggestion
    "# Now lets plot according to price.\n",
```

zoq replied:
```suggestion
    "# Lets print whichever column has missing values.\n",
```

zoq replied:
```suggestion
    " 2. For categorical data, use mode."
```

zoq replied:
```suggestion
    "'1' : If a row has that feature.<br>\n",
    "'0' : If a row doesn't have that feature."
```

zoq replied:
```suggestion
    "* Rooms per Household : To get an approximate no. of rooms each house has.\n",
    "* Bedrooms per Room : To get an approximate no. of bedrooms among total rooms.\n",
    "* Population per household : To get an approximate no. of residents in a house."
```

zoq replied:
Let's use camel casing for all names, see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#naming-conventions for more details.

zoq replied:
```suggestion
    "predictions = mlpack.linear_regression(input_model=model, test=test_set)"
```

zoq replied:
```suggestion
    "* Mean Squared Error (MSE) is calculated as the mean or average of the squared differences between predicted and expected target values in a dataset, a lower value is better.\n",
```

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,1223 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlkAAAGtCAYAAAAlE2HVAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nOy9d3hc1bX3/9lnqkajOurFklzkinvF4E43EEzvBEIJkJB2k3BvnjfJTXlzn3t/Se59k5AQCJjmACFgwgWMKcbYGBsb927LsiyrS1YZSVPP/v2xx7Zsq4zkkWyi/XmeeWZmn7332WdGmvnOWmuvJaSUaDQajUaj0Whii3GuF6DRaDQajUbzz4g ...
```
```suggestion
    "# Lets plot this on top of a piece of California map.\n",
```

swaingotnochill replied:
I am not sure what should I change here. Can you please elaborate?

zoq commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,1223 @@
+    },
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlkAAAGtCAYAAAAlE2HVAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nOy9d3hc1bX3/9lnqkajOurFklzkinvF4E43EEzvBEIJkJB2k3BvnjfJTXlzn3t/Se59k5AQCJjmACFgwgWMKcbYGBsb927LsiyrS1YZSVPP/v2xx7Zsq4zkkWyi/XmeeWZmn7332WdGmvnOWmuvJaSUaDQajUaj0Whii3GuF6DRaDQajUbzz4g ...
```
```suggestion
    "# This function will split the data into requirement ratio and shuffle so that we can as get random data.\n",
```

I think for this one we can use mlpack - https://www.mlpack.org/doc/mlpack-3.0.4/doxygen/namespacemlpack_1_1data.html#a900bb36af4ef0fff3351bb3adb09b8c5

swaingotnochill replied:
I am having difficulty in finding pre-built python functions in the docs. Is there any way to help me search the docs more effectively."OR" this utility functions can be called with the same name in python too.

swaingotnochill replied:
So, the issue is whenever I use preprocess_split(input, input_labels, ratio) method in notebook, the kernel dies...I tested it mulitple times, but it keeps on happening. So, I was thinking of sticking to this and update it when I find a solution.

zoq replied:
Do you still have the code for that one?

swaingotnochill replied:
By code do you mean the precprocess_split() method? I have that one..

zoq replied:
Yes, because I think the mlpack splitting should work, at least after the encoding.

kartikdutt18 commented on california_housing_price_prediction_with_linear_regression/California_housing_prices_predictions_with_lr_python.ipynb:
```diff
@@ -0,0 +1,1223 @@
+{
```
<p>Please add a full stop at the end of sentence.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlpack/examples/pull/158/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

kartikdutt18 replied:
<p>Please add a full stop at the end of sentence. Same in the below comments.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlpack/examples/pull/158/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

kartikdutt18 replied:
<pre class="ql-syntax" spellcheck="false">dataset.plot(
    kind="scatter",
    x="longitude",
    y="latitude", alpha=0.4,
&nbsp;&nbsp;&nbsp; s=dataset["population"] / 100,
    label="population", figsize=(10,7),
 &nbsp;&nbsp; c="median_house_value",
    cmap=plt.get_cmap("jet"),
    colorbar=True,
&nbsp;&nbsp;&nbsp;&nbsp;sharex=False
)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlpack/examples/pull/158/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

kartikdutt18 replied:
<p>Lets add space between <code>#</code>  and comment. Capitalise the comment and add full stop at the end.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlpack/examples/pull/158/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

kartikdutt18 replied:
<p>Lets a comment here that missing values are filled by median of total_bedrooms.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/mlpack/examples/pull/158/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

swaingotnochill commented:
I have added both the notebooks along with all the necessary C Python API. I also tried to resolve most the reviews, but left couple of them( i commented on those reviews). Take your time and see if this can be merged. Thanks.

P.S. - From next time, I think I should make small PRs. They are easier to track and resolve :)

zoq commented on .vscode/c_cpp_properties.json:
```diff
@@ -0,0 +1,18 @@
+{
+    "configurations": [
```
This is very specific yo your local system, so I would remove it from the PR.

swaingotnochill replied:
Yaa..it should have been added to gitignore by default..will fix this...Thanks for review.

zoq commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "C_Cpp.errorSquiggles": "Enabled"
```
See comment above.

zoq commented on student_admission_regression_with_logistic_regression/student-admission-logistic-regression-cpp.ipynb:
```diff
@@ -1,278 +0,0 @@
-{
-  "metadata":{
```
Hm, any reason to remove the file?

swaingotnochill replied:
I updated the remote from github and this created a lot of issues with my local one. Maybe at that time, this file got deleted. Also, this is a reason for "hidden files" not added to gitignore..They usually gets added to gitignore by default. I will fix this.

swaingotnochill replied:
No..I messed up.

swaingotnochill replied:
Fixing.

zoq commented on utils/heatmap.hpp:
```diff
@@ -0,0 +1,104 @@
+// Inside the C++ notebook we can use:
+// HeatMap("filename.csv",width, height,"heatmap.png")
```
I can't comment on the files directly, but looks like you added some `.pyc` files, can you remove those from the PR.

swaingotnochill replied:
Sure...I think I added them to gitignore earlier...but let me delete them...pardon

swaingotnochill replied:
My bad.

zoq commented on utils/heatmap.hpp:
```diff
@@ -0,0 +1,104 @@
+            const std::string& outFile = "histogram.png",
+            const int width = 15,
+            const int height = 10)
+
```
```suggestion
```

Remove the extra line here.

swaingotnochill commented:
The git is haunting me now..It's so easy to mess up, if you changed something in a different refs. :(

zoq commented:
> The git is haunting me now..It's so easy to mess up, if you changed something in a different refs. :(

I think you Incorporated most of the comments, what about we just close this PR and open a new one, and just make sure we don't check in the cache files?
